### DISCUSSION TITLE
External Vref

### DISCUSSION THREAD

**Discussion by NuclearPhoenixx** [earlephilhower/arduino-pico#574] [Q&A] [answered]
Hi, I'm using an external voltage reference (LM4040-3.0) and now I get more or less garbage values for temperature readings. I guess, that's because of some assumed calculations depending on 3.3V for the ADC. Any way to fix this?

**Comment by earlephilhower** [comment, marked as answer]
I think you'll need to do it using the raw Pico-SDK calls and adjust the formula accordingly. 

You can use the routine in core as a template:
https://github.com/earlephilhower/arduino-pico/blob/0afd6ae54f88b8d62fa0ff2fa371c28ea7aecf0c/cores/rp2040/wiring_analog.cpp#L140-L156

**Reply by NuclearPhoenixx** [reply]
Thank you!

**Reply by NuclearPhoenixx** [reply]
Just tried to copy this part into my Arduino sketch: 

```cpp
#include <hardware/adc.h>

float readTempCorrect() {
  adc_set_temp_sensor_enabled(true);
  delay(1); // Allow things to settle.  Without this, readings can be erratic
  adc_select_input(4); // Temperature sensor
  int v = adc_read();
  adc_set_temp_sensor_enabled(false);
  float t = 27.0f - ((v * 3.0f / 4096.0f) - 0.706f) / 0.001721f; // From the datasheet
  return t;
}
```
And it compiled without any errors, however, now it reads ~ 80 °C instead of the correct ~ 27 °C. Any ideas?

**Reply by earlephilhower** [reply]
Did you call `adc_init()` somewhere prior?  I'm not sure if there is add'l startup time needed on the 1st init (like when changing inputs).

You could also check forums.raspberryoi.com since you're really just using the raw SDK here.  They might have some tricks or have done it already.

**Reply by NuclearPhoenixx** [reply]
Oh, actually it was completely my fault. I was reading analog values in an ISR which seems to have interfered with the ADC calls in this function. Don't know why this never happened when using the `analogReadTemp()` method directly though.
